### ISSUE TITLE
"Open in default browser" opening wrong browser

### ISSUE THREAD

**Issue by henderea** [dmarmor/epichrome#114]
I have an epichrome app for gmail, and I have it set so that certain work-related links will open in my default browser, which is Chrome.  All of a sudden, it's opening up Firefox.  And if I delete Firefox.app, it opens Safari.  The only thing I can think of that changed other than maybe the latest Chrome update was that I upgraded macOS from 10.12.4 to 10.12.5 last night.  I've tried using Cocktail to rebuild the launch services database (though without selecting the reset file associations option), and that didn't help any.  Also, when I try to set Chrome as the default browser, the system dialog that pops up asking if I want to keep the current default or use the new one shows one of my epichrome apps as the name instead of Chrome.

Setting the default browser to browser fairy (which i had been using before) doesn't help any.  Same with setting it to Safari: even with the default browser as safari, the link *still* opens in Firefox.  Is there any way I can clear the cache or whatever that Chrome/Epichrome has for default browser?

I should probably point out that **the issue does not happen with non-epichrome apps, like Slack or others.**  They open the correct default browser.  This seems to be an issue either with that specific epichrome app, or all of my epichrome apps (not sure which, since I don't have my other epichrome apps set up like that).

To try and fix the issue, I have made a few different attempts at resetting things related to the browsers.  One thing I tried was reinstalling chrome, deleting the .app and putting a freshly downloaded one in its place.  Another was using CleanMyMac to reset virtually all of the application data for chrome, leaving little more than the .app, dock icon, and application support folder.  On top of that, I've attempted to delete and re-create my Gmail epichrome app (without registering it as a browser this time), however in order to not lose data, I copied back the contents of the application support folder after recreating it.  The last thing I tried was exporting my epichrome rules, deleting the extension data folder for epichrome helper in that profile, and starting the gmail app back up, re-importing the rules.  None of these things worked.

**Comment by asos-rshev** [comment]
The same is happening to me after updating to 10.12.5. Opens Safari regardless of any settings. Extremely annoying.

**Comment by GrogHills** [comment]
Same as me.

**Comment by henderea** [comment]
Actually, I don't have firefox set up for anything in browser fairy.  I have one work-related site going to a separate epichrome app, and everything else goes to regular chrome.  I don't really use anything other than chrome for regular browsing; firefox is basically just installed so i can test webpages in it if needed.  And firefox has never been my default browser on any computer I've ever owned, so it's not like it dug up some really old setting.  It just doesn't want to use chrome, and epichrome apps are lumped in with that.

I don't really have any idea what happened to cause the issue all of a sudden, but it seems like the epichrome apps suddenly don't recognize any chrome-based browser as a default browser, which I guess means they're using some sort of fallback logic.

It may actually be an issue with the epichrome helper *extension*, since that's what manages the rules for opening in other browsers.  Maybe 10.12.5 or a recent chrome update changed something with how extensions open default apps for a file type, and the epichrome extension isn't handling the change well.

**Comment by henderea** [comment]
Hey, I found a solution, though it's slightly technical.

**Step 1:** install the latest python (2.7.13) with homebrew.  if you don't understand this step, it may be too technical for you.  If you want to continue anyway, see the end of this comment for more info.
**Step 2:** open up `/Users/<your username>/Library/Application Support/Epichrome/Apps/<app name>/NativeMessagingHosts/org.epichrome.helper-host.py` in a text/code editor and change the first line of the file from
```#!/usr/bin/python```
to
```#!/usr/local/bin/python```
so that it uses the homebrew version of python.
**Step 3:** quit the epichrome app (if you left it open) and relaunch it.  It should now be working.

Here is what lead me to this discovery:
I ran `/usr/bin/python -V` to get the version of the built-in python.  It returned `2.7.10`.  Looking around for python bugs related to the webbrowser module lead me to <http://bugs.python.org/issue24452>, which was a bug where the webbrowser module didn't properly support chrome.  It looks to me like the fix got released with Python 2.7.13 (according to <https://hg.python.org/cpython/raw-file/v2.7.13/Misc/NEWS>), so mac comes with a broken version.  The latest python 2.7 is 2.7.13, and that's what homebrew installs.  Using the homebrew installation fixes the issue.

I couldn't find any info on what version of python came with 10.12.4, but searching on google seems to indicate that 10.11 El Capitan came with python 2.7.10, so I'm honestly not sure why it worked before.

### Installing python with homebrew
If you aren't familiar with homebrew, it is a package manager for OS X/macOS.  Basically, it lets you install command-line stuff like python or ruby.  If you don't already have homebrew set up, you can run the command from <https://brew.sh/> to install it.  After homebrew is installed, run the following commands:
```brew update```
```brew install python```
If you get any errors, google for them. homebrew has a lot of users, so you'll probably find it difficult to discover an issue that no one else has encountered.
And that's it, python is now installed with homebrew.  To double-check that it worked, run
```python -V```
in the terminal and make sure it gives `2.7.13`.  Please note that it needs to be an UPPERCASE `V`.  Lowercase `v` means something completely different, so if you get a bunch of stuff printed out and a unusual prompt, you did a lowercase v and will need to run the `exit()` command to get out of the python command prompt.  I made that mistake myself 🙂

**Comment by GrogHills** [comment]
You are a star. Thank you so much! I just followed your amazing, detailed, well-written directions to make the changes and it definitely fixed the issue.

**Comment by henderea** [comment]
Glad it helped.  TBH, I actually spent 30+ minutes looking into the Epichrome code and trying to figure things out when I was *supposed* to be working on work-related code 🙂 It turned out to be a simple change, but I went through all of the epichrome extension code looking for how it opened the default browser, which eventually lead me to discover that it uses a python script to handle that.

Oh, and if you want to apply this to future epichrome apps, you can make that one-line change in `/Applications/Epichrome.app/Contents/Resources/Runtime/Resources/NativeMessagingHosts/org.epichrome.helper-host.py`.  If you're using finder, go to the Applications folder and right click on the Epichrome app, then select the "show package contents" option to get inside the app.  I've always thought it was cool that mac apps are actually just special folders 🙂  If you want to make sure the change will stick around for an already-created app, edit the `/Applications/<your app>.app/Contents/Resources/NativeMessagingHosts/org.epichrome.helper-host.py` file, making the same 1-line change.

**Comment by henderea** [comment]
I've created a script that _should_ automate updating of the epichrome apps to use homebrew python.  You'll have to install python 2.7.13 with homebrew first, but then you can run the script in [this gist](https://gist.github.com/henderea/57a83deb39785e3be0890418b5b85790) to update the files that my earlier instructions had you doing manually.  Please note that some of this uses the `locate` command to find copies of the file, so if you have spotlight turned off, that might affect things.  Also, if an app was just created or moved recently, it won't pick up on that.  But as long as you haven't made new apps or moved apps recently, you'll probably be just fine.

To run it, download the file from the Gist and run it with ```bash switch-epichrome-python.sh``` from whatever directory you have the file in.

**Comment by bold-joe-ocampo** [comment]
@henderea you rock!

**Comment by asos-rshev** [comment]
@henderea That's strange, I have 2.7.13 Python bundled with OSX 10.12.5:
> $ python -V
Python 2.7.13
$ whereis python
/usr/bin/python

Doubt there would be a point to install homebrew python with the same version... Will post an update here.

**Comment by asos-rshev** [comment]
Strange but it helped. Something's terribly wrong with Apple's Python, despite the same version number.

**Comment by purevertigo** [comment]
@dmarmor Is there an ETA for this fix? I am not savvy with manual fixes.

thanks

**Comment by henderea** [comment]
@purevertigo:
If my manual instructions are a bit too much for you, you might try this simplified approach I set up:

Open up `/Applications/Utilities/Terminal.app`
Once it opens, paste in this command (make sure you get the whole line) and hit enter:
```bash
curl -sSL 'https://gist.githubusercontent.com/henderea/57a83deb39785e3be0890418b5b85790/raw/a39d2b7696b1ee342093b5624e29b02a24bfdc19/switch-epichrome-python.sh' | bash
```

You may have to fill in some prompts, and maybe your admin password at some point during the setup, but it should have stuff mostly automated.

Let me know if that command works for you.

**Comment by purevertigo** [comment]
@henderea confirmed, two steps and now works great. (I did have to reboot though).

thanks!

**Comment by breeden** [comment]
@henderea from other threads I get the impression @dmarmor is busy with his real job these days. you should totally branch and submit a pull request!

**Comment by zzamboni** [comment]
@henderea thank you for identifying the issue - I had the same problem and it had been driving me crazy. Your fix stops SSBs from force-opening Firefox or Safari, but unfortunately it still doesn't send the URLs to my default browser (Hammerspoon, from which I do URL dispatching to different SSBs). But your comments have put me on the right track, and I'll keep investigating from here.

**Comment by henderea** [comment]
I'll make an attempt at a pull request.

**Comment by henderea** [comment]
Pull request #115 is my fix for the issue.  You can probably apply those changes manually to the python file I identified earlier if you need a browser not supported by the existing code even with the latest python.

**Comment by henderea** [comment]
@dmarmor: If you could get my pull request merged and a new version released, that would be really helpful to a lot of people.  If there are additional changes needed, such as for updating the file in the existing epichrome apps, let me know what needs to be done, and I can give it a shot if you don't have much spare time.

**Comment by samlimster** [comment]
Hi @henderea 

Thanks for all your great work on this issue.  I tried your script:
1) installed python
2) downloaded your script to my desktop
3) ran it.  It said that I needed to run a command to update some parameters so I ran that.
4) ran your script again and this time it completed successfully.
5) closed and reopened my ephichrome apps

I have browser fairy set as the default browser

Prior to running your script, all epichrome links opened in Safari which was annoying.  However, after running your script, all epichrome links open in Chrome - even though Browser fairy is my default browser.

Did I miss something?

Thanks @henderea

**Comment by henderea** [comment]
@samlimster: you didn't miss anything.  My script was just to fix the problem opening chrome.  Take a look at my pull request #115 to see the better fix.  You would apply those changes in the same files that the script said it updated (the `org.epichrome.helper-host.py` files).  Unfortunately, there's not a good way to automate that fix, so we'll have to wait for a release to get it applied automatically.

**Comment by samlimster** [comment]
Thanks @henderea !  This worked like a charm! :)

**Comment by justone** [comment]
Thanks @henderea for your fix.  Editing the generated apps didn't work for me, but editing Epichrome directly and regenerating the apps did.  Thanks again!

**Comment by henderea** [comment]
I created a new [script](https://gist.github.com/henderea/9658db4a82b1067d6ff8b08cf7683910) to automate applying the changes from #115 to your existing and future apps

```bash
curl -sSL 'https://gist.githubusercontent.com/henderea/9658db4a82b1067d6ff8b08cf7683910/raw/85843029f53b641d32851c2042c3874ce96343d7/apply-epichrome-patch.sh' | bash
```

Let me know how that works.

**Comment by andresrinivasan** [comment]
@henderea Are you assuming GNU sed?  Tried running the script and got `sed: 1: "s/(\s*)import webbrowse ...": \1 not defined in the RE`

**Comment by henderea** [comment]
I just used whatever sed I had installed.  Must have been the homebrew one.  The script has been updated in my previous comment.  Luckily, I had already done scripting for setting stuff up with homebrew in my earlier python upgrade script 🙂

**Comment by wizonesolutions** [comment]
@henderea I used the updated script, but I still got:

```
Updating Epichrome files in existing apps
sed: 1: "s/(\s*)import webbrowse ...": \1 not defined in the RE
```

I think it's because Homebrew aliases `sed` to `gsed`.

**Comment by wizonesolutions** [comment]
That worked now.

**Comment by dpickett** [comment]
@henderea you rock - thank you for this patch. Worked like a charm for me. The script was super helpful.

**Comment by santry** [comment]
@asos-rshev it sounds like you already have the Homebrew version of python 2.7.13 installed. The `whereis` command just finds programs on your system. It doesn't tell you which one is in the user's path.

To check which version of python is being used from the user's path, you want to use `which python`. I'll bet that returns `/usr/local/bin/python`, which is the Homebrew version.

To explicitly check the OS X installed version, use `/usr/bin/python -V`, which should be `2.7.10`.

**Comment by matthewfallshaw** [comment]
Without requiring sed to be installed `--with-default-names`:

    curl -sSL 'https://gist.githubusercontent.com/matthewfallshaw/35fe2b89122e9247b29ae678e0847090/raw/b77265ea8a702b3d376c8b01522e0c3a5c5742a9/apply-epichrome-patch.sh' | bash

Thank you, @henderea.

**Comment by kota77** [comment]
Thank you so much, @henderea! You rock!

**Comment by jtbayly** [comment]
Yep. Thanks!

**Comment by dmarmor** [comment]
I've integrated @henderea's fix in 2.1.17 which should be up in the next hour or so. Thank you all for the help tracking this down and fixing it!

**Comment by rgbink** [comment]
Still an issue. Applied @henderea 's fix. Fri, Jan 11, 2019. ¯\_(ツ)_/¯
